### ISSUE TITLE
Export Data Error from Database (django.db.utils.OperationalError: too many SQL variables)

### ISSUE THREAD

**Issue by maxstrobel** [heartexlabs/label-studio#3109]
Hi,

I found already #1941, which contains basically the same issue.

I've ~3.5k images in my database and I'm getting now the same error when trying to export data:

~~~
Traceback (most recent call last):
  File "/opt/app-root/lib64/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/opt/app-root/lib64/python3.9/site-packages/django/db/backends/sqlite3/base.py", line 423, in execute
    return Database.Cursor.execute(self, query, params)
sqlite3.OperationalError: too many SQL variables

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/app-root/lib64/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/opt/app-root/lib64/python3.9/site-packages/django/utils/decorators.py", line 43, in _wrapper
    return bound_method(*args, **kwargs)
  File "/label-studio/label_studio/data_export/api.py", line 177, in get
    tasks += ExportDataSerializer(
  File "/opt/app-root/lib64/python3.9/site-packages/rest_framework/serializers.py", line 768, in data
    ret = super().data
  File "/opt/app-root/lib64/python3.9/site-packages/rest_framework/serializers.py", line 253, in data
    self._data = self.to_representation(self.instance)
  File "/opt/app-root/lib64/python3.9/site-packages/rest_framework/serializers.py", line 686, in to_representation
    return [
  File "/opt/app-root/lib64/python3.9/site-packages/django/db/models/query.py", line 280, in __iter__
    self._fetch_all()
  File "/opt/app-root/lib64/python3.9/site-packages/django/db/models/query.py", line 1324, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/opt/app-root/lib64/python3.9/site-packages/django/db/models/query.py", line 51, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/opt/app-root/lib64/python3.9/site-packages/django/db/models/sql/compiler.py", line 1175, in execute_sql
    cursor.execute(sql, params)
  File "/opt/app-root/lib64/python3.9/site-packages/sentry_sdk/integrations/django/__init__.py", line 560, in execute
    return real_execute(self, sql, params)
  File "/opt/app-root/lib64/python3.9/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/opt/app-root/lib64/python3.9/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/opt/app-root/lib64/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/opt/app-root/lib64/python3.9/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/opt/app-root/lib64/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/opt/app-root/lib64/python3.9/site-packages/django/db/backends/sqlite3/base.py", line 423, in execute
    return Database.Cursor.execute(self, query, params)
django.db.utils.OperationalError: too many SQL variables
~~~

My versions:
~~~
{
  "release": "1.5.0dev",
  "label-studio-os-package": {
    "version": "1.5.0dev",
    "short_version": "1.5",
    "latest_version_from_pypi": "1.6.0",
    "latest_version_upload_time": "2022-09-27T17:09:34",
    "current_version_is_outdated": true
  },

  "label-studio-os-backend": {},

  "label-studio-frontend": {
    "message": "fix: DEV-3061: Image scaling goes lower than 1 causing small images (# ...",
    "commit": "10dd11839b96d07f8978270896dcd76464c06cc4",
    "branch": "master",
    "date": "2022-08-29T18:16:33Z"
  },

  "dm2": {
    "message": "fix: DEV-2519: Choices selected by default",
    "commit": "dc03b7337518a3129677570127b44456420d4c6c",
    "branch": "master",
    "date": "2022/08/10 15:51:21"
  },

  "label-studio-converter": {
    "version": "0.0.44"
  }
}
~~~

Now I've a few questions:

1. From the documentation (https://labelstud.io/guide/storedata.html#Labeling-performance):

> The SQLite database works well for projects with tens of thousands of labeling tasks. 

Am I doing something wrong, or why does this happen already for 3.5k images?

2. @makseq's workaround to export data (https://github.com/heartexlabs/label-studio/issues/1941#issuecomment-1121609239) works - why do I not get any error here? It's the same database in the end.
3. Is there a way to migrate the SQLite database to PostgreSQL?

**Comment by makseq** [comment]
Hi, 

> The SQLite database works well for projects with tens of thousands of labeling tasks.

Good point, I've rewritten this paragraph. 

> Am I doing something wrong, or why does this happen already for 3.5k images?

Most likely you are using data manager filters or another logical things that make SQL queries complex. 

> why do I not get any error here? It's the same database in the end.

Because the result SQL query becomes much easier. 

> Is there a way to migrate the SQLite database to PostgreSQL?

Unfortunately not yet. However, you can export data from the project and re-import it on LS with the PostgreSQL instance.

**Comment by maxstrobel** [comment]
Are there guides how to export & re-import the data?

I found some fuzzy descriptions how to do this for the labels & data itself, but, for example, what about the users?

**Comment by makseq** [comment]
Good idea about a guide for this.. 
You need to recreate the same users (with the same emails) manually or via API / command line.

**Comment by nachouve** [comment]
Below is a workaround I employed to address the issue:

I encountered a similar issue on a machine where I lack administrative privileges, preventing me from updating software. The error stemmed from my SQLite version, specifically 3.26.0, which faced constraints related to the SQLITE_MAX_VARIABLE_NUMBER limit. 

I addressed this by adjusting the batch size to 500 [here](https://github.com/HumanSignal/label-studio/blob/efa1e7b7ac04f7a6b3c989732667fd5e930af8f2/label_studio/data_export/api.py#L191), successfully exporting all files.
